Python/optimizer.c

```diff
@@ -154,15 +155,19 @@ _PyOptimizer_BackEdge(_PyInterpreterFrame *frame, _Py_CODEUNIT *src, _Py_CODEUNI
         return frame;
     }
     _PyOptimizerObject *opt = interp->optimizer;
-    _PyExecutorObject *executor = opt->optimize(opt, frame->f_code, dest);
-    if (executor == NULL) {
-        return NULL;
+    _PyExecutorObject *executor;
+    int err = opt->optimize(opt, frame->f_code, dest, &executor);
+    if (err <= 0) {
+        if (err < 0) {
+            return NULL;
+        }
+        _PyFrame_SetStackPointer(frame, stack_pointer);
+        return frame;
```

Comment:
We could also complicate the signature of `_PyOptimizer_BackEdge` to signal this case back to the caller so we could avoid the stack pointer write, and the caller could just continue regular interpretation instead of needing to `goto resume_frame` and update interpreter state from the frame again. But probably not worth it I guess, if optimizers declining to optimize should be rare?

Reply:
Most calls to `_PyOptimizer_BackEdge` will return the same frame. We could add a check for it in `JUMP_BACKWARD`, but I think we would still need to save the stack pointer. 
Even if we could avoid it, I suspect a single unconditional write is faster than the additional logic to avoid it.